db/system_keyspace.cc

```diff
@@ -45,6 +45,7 @@
 #include <boost/range/adaptor/map.hpp>
 
 #include <seastar/core/coroutine.hh>
+#include <seastar/core/reactor.hh>
```

Comment:
We might add a system.metrics vtable later to query all metrics.

Reply:
I agree that this would be a good candidate. These kinds of virtual tables will make development so much easier. I often want to check certain metrics but don't feel like firing up monitoring for my 2 minutes quick test, so I just work around it instead.

Reply:
Rgarding the metrics. There's a `collectd` gate in seastar that, AFAIK, was supposed to be used as a lightweight replacement for graphana. I tried it several times with no luck, so I'm all for having a _working_ method to work with metrics. But we already have one -- the prometheus exporter itself. It's just missing some handy collecting/accepting/reading piece. Maybe having a lightweight "prometheus server" would work better then system.metrics vtable?

Reply:
There are multiple problems with using the prometheus protocol for ad-hoc metric reading:
* Individual scrapes only contain changed values (not 100% about this);
* Horrible, bloated text-based format;
* No way to filter for a specific metric;
* A scrape is significant effort (not really an issue for the dev use-case I mentioned above);

Since the metrics layer contains all the metadata, a vtable would be quite easy to write, somewhat similar to the config table you wrote.

Reply:
> Individual scrapes only contain changed values (not 100% about this);

Me neither, but I've a script that prints IO metrics in `vmstat 1` manner and I never made any workarounds for missing values, so likely it emits everything.

> Horrible, bloated text-based format;

Agree, but I'm biased here (parsing Linux /proc files is not much less horrible)

> No way to filter for a specific metric;

Well, no way to filter it server-side, so you scrape first, filter next. But in case of vtable you don't filter it by name on scylla side either, do you?

> A scrape is significant effort (not really an issue for the dev use-case I mentioned above);

It's 2-fold. First you collect them, second you send them via http. In case of vtable the first part doesn't go away (or does it?), so the choise is between http vs cql, no?

Reply:
> > Individual scrapes only contain changed values (not 100% about this);
> 
> Me neither, but I've a script that prints IO metrics in `vmstat 1` manner and I never made any workarounds for missing values, so likely it emits everything.

Maybe we only wanted to do it but didn't get around to. I only remember that we wanted to reduce the strain scrapes put on the system and this idea was floated at one point, we might not have gotten around to actually implement it. Or maybe the idea was about internally only "re-rendering" parts of the text which was changed.

> 
> > Horrible, bloated text-based format;
> 
> Agree, but I'm biased here (parsing Linux /proc files is not much less horrible)

At one point I had a python script which could get metric values, but it might have been using the REST API actually.

> 
> > No way to filter for a specific metric;
> 
> Well, no way to filter it server-side, so you scrape first, filter next. But in case of vtable you don't filter it by name on scylla side either, do you?

If you set up the table schema right, filtering metrics comes down to selecting a partition (or even down to a clustering row), which is then propagated to the vtable as the read range.

> 
> > A scrape is significant effort (not really an issue for the dev use-case I mentioned above);
> 
> It's 2-fold. First you collect them, second you send them via http. In case of vtable the first part doesn't go away (or does it?), so the choise is between http vs cql, no?

I guess so, also with more efficient filtering.

Reply:
There is scyllatop that talk with the prometheus protocol and displays metrics. Really, metrics are the least interesting virtual table because there is already plenty of infrastructure around metrics. It's nice to have for developers but production use doesn't really need it. We should provide it if it's easy, but no need to make a large effort around it since it won't be commonly used.